### PR TITLE
Feature PCA outlier detector

### PR DESCRIPTION
**Pull request by mauicv**
## What is this?

This PR adds the PCA outlier detector. See a simple example in this [notebook](https://colab.research.google.com/gist/mauicv/6ca5ac54a0f997bcbd7b0c839272df0c/pca-detector-pr728ipynb.ipynb). This PR branches off the [mahalanobis outlier detector work](https://seldondev.slack.com/archives/CJJ7E0BNY/p1674486960826709).

## TODO:

- [x] Add more here
- [x] Example
- [x] tests
- [x] optional-deps
- [x] torch-scriptable
- [x] docstrings
- [x] Merge kNN `fit`/`infer_threshold` changes

### CODE REVIEW

File: alibi_detect/od/_pca.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        likely to be an outlier.
+
+        When scoring a test instance we project it onto the eigenvectors and compute its score using the L2 norm. If
+        a threshold is fitted we use this to determine whether the instance is an outlier or not.
+
+        Parameters
+        ----------
+        n_components:
+            The number of dimensions in the principle subspace. For linear pca should have
+            ``1 <= n_components < dim(data)``. For kernel pca should have ``1 <= n_components < len(data)``.
```
Comment (ascillitoe) [inline alibi_detect/od/_pca.py:48]: Since `n_components` is pretty fundamental to the algo, do we want to add a check and clear error message for the  `n_components >= dim(data)` case?

File: alibi_detect/od/_pca.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        Raises
+        ------
+        NotFittedError
+            If called before detector has been fit.
+        """
+        score = self.backend.score(self.backend._to_tensor(x))
+        return self.backend._to_numpy(score)
+
+    @catch_error('NotFittedError')
+    def infer_threshold(self, x_ref: np.ndarray, fpr: float) -> None:
```
Comment (ojcobb) [inline alibi_detect/od/_pca.py:124]: x_ref -> x for consistency with previous detectors

File: alibi_detect/od/_pca.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        """Infer the threshold for the kNN detector.
```
Comment (ojcobb) [inline alibi_detect/od/_pca.py:125]: knn -> pca

File: alibi_detect/od/pytorch/pca.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+    def fit(self, x_ref: torch.Tensor) -> None:
+        """Fits the PCA detector.
+
+        Parameters
+        ----------
+        x_ref
+            The Dataset tensor.
+        """
+        self.x_ref_mean = x_ref.mean(0)
```
Comment (ojcobb) [inline alibi_detect/od/pytorch/pca.py:82]: Does this only need computing for the linear variant?

File: alibi_detect/od/pytorch/pca.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        self.pcs = self.compute_pcs(x_ref)
+        self.x_ref = x_ref
```
Comment (ojcobb) [inline alibi_detect/od/pytorch/pca.py:84]: Is it necessary to store `x_ref`?

File: alibi_detect/od/tests/test__knn/test__knn.py [hunk 6]
```diff
@@ -244,7 +260,12 @@ def test_knn_integration():
...
     assert result
 
-    tsknn = torch.jit.script(knn_detector.backend)
+    ts_knn = torch.jit.script(knn_detector.backend)
     x = torch.tensor([x_inlier[0], x_outlier[0]], dtype=torch.float32)
-    y = tsknn(x)
+    y = ts_knn(x)
+    assert torch.all(y == torch.tensor([False, True]))
+
+    ts_knn.save(tmp_path / 'pca.pt')
```
Comment (ojcobb) [inline alibi_detect/od/tests/test__knn/test__knn.py:268]: I guess it doesn't make a difference, but should `pca.pt` really be `knn.pt` here? Likewise for Mahalanobis test

File: alibi_detect/od/pytorch/pca.py (hunk 5 above)
Comment (ojcobb) [reply alibi_detect/od/pytorch/pca.py:84]: (For linear variant)

File: alibi_detect/od/pytorch/pca.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            The test data.
+
+        Returns
+        -------
+        The outlier score.
+        """
+        k_xr = self.kernel(x, self.x_ref)
+        k_xr_row_sums = k_xr.sum(1)
+        _, n = k_xr.shape
+        k_xr_cen = k_xr - self.k_col_sums[None, :]/n - k_xr_row_sums[:, None]/n + self.k_sum/(n**2)
```
Comment (ojcobb) [inline alibi_detect/od/pytorch/pca.py:214]: This should be

m, n = k_xr.shape
k_xr_cen = k_xr - self.k_col_sums[None, :]/m - k_xr_row_sums[:, None]/n + self.k_sum/(m*n)

My bad - I see this was a typo in my original demo

**Comment by ojcobb** [comment]
Nice work! A few minor comments but otherwise LGTM!

File: alibi_detect/od/_pca.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        backend: Literal['pytorch'] = 'pytorch',
+        device: Optional[Union[Literal['cuda', 'gpu', 'cpu'], 'torch.device']] = None,
+    ) -> None:
+        """Principal Component Analysis (PCA) outlier detector.
+
+        The detector is based on the Principal Component Analysis (PCA) algorithm. There are two variants of PCA:
+        linear PCA and kernel PCA. Linear PCA computes the eigenvectors of the covariance matrix of the data. Kernel
+        PCA computes the eigenvectors of the kernel matrix of the data. In each case, we choose the smallest
+        `n_components` eigenvectors. We do this as they correspond to the invariant directions of the data. i.e the
+        directions along which the data is least spread out. Thus a point that deviates along these dimensions is more
```
Comment (ascillitoe) [inline alibi_detect/od/_pca.py:38]: Extreme nitpick. Starting a sentence with `i.e.` is odd!

File: alibi_detect/od/_pca.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    'pytorch': (KernelPCATorch, LinearPCATorch)
+}
+
+
+class PCA(BaseDetector, ThresholdMixin, FitMixin):
+    def __init__(
+        self,
+        n_components: int,
+        kernel: Optional[Callable] = None,
+        backend: Literal['pytorch'] = 'pytorch',
```
Comment (ascillitoe) [inline alibi_detect/od/_pca.py:29]: Is the itention also to provide a `keops` backend in the future? Like with Mahalonobis...

File: alibi_detect/od/_pca.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        device: Optional[Union[Literal['cuda', 'gpu', 'cpu'], 'torch.device']] = None,
```
Comment (ascillitoe) [inline alibi_detect/od/_pca.py:30]: The `device` type is getting pretty long, is it worth defining it in `_types.py`?

File: alibi_detect/od/tests/test__knn/test__knn.py [hunk 11]
```diff
@@ -24,6 +24,17 @@ def make_knn_detector(k=5, aggregator=None, normalizer=None):
 def test_unfitted_knn_single_score():
     knn_detector = KNN(k=10)
     x = np.array([[0, 10], [0.1, 0]])
+    x_ref = np.random.randn(100, 2)
```
Comment (ascillitoe) [inline alibi_detect/od/tests/test__knn/test__knn.py:27]: Are these tests being added here to sync the knn and mahalanobis tests with those used for the pca detector? Wondering if it might be worth opening a separate PR for that, or renaming this one slightly. Just so reverting this one in the future wouldn't have the unintended effect of changing the knn and mahalanonbis tests.

File: alibi_detect/od/pytorch/pca.py (hunk 5 above)
Comment (mauicv) [reply alibi_detect/od/pytorch/pca.py]: Yeah, good point, I've moved to the kernelPCA class!

File: alibi_detect/od/pytorch/pca.py (hunk 7 above)
Comment (mauicv) [reply alibi_detect/od/pytorch/pca.py]: Ah, good point. Although I think we've got `m` and `n` the wrong way around above.

File: alibi_detect/od/_pca.py (hunk 9 above)
Comment (mauicv) [reply alibi_detect/od/_pca.py:29]: Yep

File: alibi_detect/od/pytorch/pca.py (hunk 7 above)
Comment (ojcobb) [reply alibi_detect/od/pytorch/pca.py]: Yes, good spot!

File: alibi_detect/od/_pca.py (hunk 10 above)
Comment (mauicv) [reply alibi_detect/od/_pca.py:30]: Good idea, I might defer this to a future PR though as probably makes sense to make this change in other places too!

Comment (mauicv) [reply alibi_detect/od/_pca.py:30]: I'll open an issue

**Review by ojcobb — APPROVED** [review]

File: alibi_detect/od/_pca.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    def fit(self, x_ref: np.ndarray) -> None:
+        """Fit the detector on reference data.
+
+        In the linear case we compute the principle components of the reference data using the
+        covariance matrix and then remove the largest `n_components` eigenvectors. The remaining
+        eigenvectors correspond to the invariant dimensions of the data. Changes in these
+        dimensions are used to compute the outlier score which is the distance to the principle
+        subspace spanned by the first `n_components` eigenvectors.
+
+        In the kernel Case we compute the principle components of the reference data using the
```
Comment (ascillitoe) [inline alibi_detect/od/_pca.py:99]: `kernel Case` -> `kernel case`

File: alibi_detect/od/_pca.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+        kernel matrix and then return the largest `n_components` eigenvectors. These are then
+        normalized to have length equal to `1/eigenvalue`. Note that this differs from the
+        linear case where we remove the largest eigenvectors.
+
+        In both cases we then store the computed components and use later when we score test
```
Comment (ascillitoe) [inline alibi_detect/od/_pca.py:104]: `and use later` -> `to use later`?

File: alibi_detect/od/pytorch/pca.py [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+from alibi_detect.od.pytorch.base import TorchOutlierDetector
+
+
+class PCATorch(TorchOutlierDetector):
+    ensemble = False
+
+    def __init__(
+            self,
+            n_components: int,
+            device: Optional[Union[str, torch.device]] = None
```
[the remark this answers is not in the document]
Comment (ascillitoe) [reply alibi_detect/od/pytorch/pca.py:14]: Noting that the same pattern is used in the KNN detector, so if there isn't a reason, seems reasonable to add a note about this in #779...

File: alibi_detect/od/pytorch/pca.py [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+        Raises
+        ------
+        ThresholdNotInferredException
+            If called before detector has had `infer_threshold` method called.
+        """
+        scores = self.score(x)
+        if not torch.jit.is_scripting():
+            self.check_threshold_inferred()
+        preds = scores > self.threshold
+        return preds.cpu()
```
Comment (ascillitoe) [inline alibi_detect/od/pytorch/pca.py:59]: Is the `.cpu()` needed here? Since isn't it already done in `.to_numpy()`?

File: alibi_detect/od/pytorch/pca.py [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+        Tensor of scores for each element in `x`.
+
+        Raises
+        ------
+        NotFitException
+            If called before detector has been fit.
+        """
+        self.check_fitted()
+        score = self.compute_score(x)
+        return score.cpu()
```
Comment (ascillitoe) [inline alibi_detect/od/pytorch/pca.py:80]: Is the .cpu() needed here? Since isn't it already done in .to_numpy()?

File: alibi_detect/od/tests/test__knn/test__knn.py (hunk 11 above)
Comment (ascillitoe) [reply alibi_detect/od/tests/test__knn/test__knn.py:27]: OK fair enough, I don't feel that strongly!

File: alibi_detect/od/pytorch/pca.py [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+            self,
+            n_components: int,
+            device: Optional[Union[str, torch.device]] = None
+            ):
+        """Linear variant of the PyTorch backend for PCA detector.
+
+        Parameters
+        ----------
+        n_components:
+            The number of dimensions in the principle subspace.
```
Comment (ascillitoe) [inline alibi_detect/od/pytorch/pca.py:111]: `principle` or `principal`?

File: alibi_detect/od/pytorch/pca.py (hunk 16 above)
Comment (mauicv) [reply alibi_detect/od/pytorch/pca.py:80]: GOod catch! :+1:

File: alibi_detect/od/pytorch/pca.py (hunk 17 above)
Comment (mauicv) [reply alibi_detect/od/pytorch/pca.py]: Yeah your right, it should be `principal`

**Review by ascillitoe — APPROVED** [review]
LGTM!
